src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,681 @@
+	if (!video_encoder.is_initalised) {
+		if (!video_encoder.init(image.width,
+		                        image.height,
+		                        image.pixel_format,
+		                        rounded_fps)) {
+			LOG_ERR("FFmpeg: Video encoder failed to init");
+			return;
+		}
+		bool init_muxer = false;
+		if (!muxer.is_initalised && audio_encoder.is_initalised) {
+			init_muxer = muxer.init(video_encoder, audio_encoder);
+			if (!init_muxer) {
+				LOG_ERR("FFmpeg: Failed to init muxer");
+				video_encoder.free();
+				return;
+			}
+		}
+		mutex.lock();
+		video_encoder.is_initalised = true;
+		muxer.is_initalised         = muxer.is_initalised || init_muxer;
+		mutex.unlock();
+		waiter.notify_all();
+	} else if (video_encoder.width != image.width ||
+	           video_encoder.height != image.height ||
+	           video_encoder.pixel_format != image.pixel_format ||
+	           video_encoder.frames_per_second != rounded_fps) {
+		capture_video_finalise();
+		if (!video_encoder.init(image.width,
+		                        image.height,
+		                        image.pixel_format,
+		                        rounded_fps)) {
+			LOG_ERR("FFmpeg: Video encoder failed to init");
+			return;
+		}
+		mutex.lock();
+		video_encoder.is_initalised = true;
+		mutex.unlock();
+		waiter.notify_all();
+	}
+	RenderedImage copy = image.deep_copy();
+	if (!video_encoder.queue.Enqueue(std::move(copy))) {
+		// Should never happen but avoid a memory leak if it does
+		LOG_ERR("FFmpeg: Video encoder queue is stopped");
+		copy.free();
+	}
+}
+
+void FFmpegEncoder::capture_video_add_audio_data(const uint32_t sample_rate,
+                                                 const uint32_t num_sample_frames,
+                                                 const int16_t* sample_frames)
+{
+	if (!audio_encoder.is_initalised) {
+		if (!audio_encoder.init(sample_rate)) {
+			LOG_ERR("FFmpeg: Failed to init audio encoder");
+			return;
+		}
+		bool init_muxer = false;
+		if (!muxer.is_initalised && video_encoder.is_initalised) {
+			init_muxer = muxer.init(video_encoder, audio_encoder);
+			if (!init_muxer) {
+				LOG_ERR("FFmpeg: Failed to init muxer");
+				audio_encoder.free();
+				return;
+			}
+		}
+		mutex.lock();
+		audio_encoder.is_initalised = true;
+		muxer.is_initalised         = muxer.is_initalised || init_muxer;
+		mutex.unlock();
+		waiter.notify_all();
+	} else if (audio_encoder.sample_rate != sample_rate) {
+		capture_video_finalise();
+		if (!audio_encoder.init(sample_rate)) {
+			LOG_ERR("FFmpeg: Failed to init audio encoder");
+			return;
+		}
+		mutex.lock();
+		audio_encoder.is_initalised = true;
+		mutex.unlock();
+		waiter.notify_all();
+	}
+
+	// Thought about writing a BulkEnqueue method that takes a pointer
+	// To avoid the temporary std::vector
+	// But this entire function only takes 5 microseconds average on my
+	// computer and removing the std::vector didn't help it that much
+	const size_t num_frames = num_sample_frames * 2;
```

Comment:
How come? Did you want to write `num_samples =`? Even then, `num_sample_frames * SamplesPerFrame` would make this clearer.

Reply:
Changed to num_samples and extracted the SamplesPerFrame constant